OpenAPI/Swagger JSON type schema support + many small fixes I came across

Jolanrensen opened the pull request:
As described by https://github.com/Kotlin/dataframe/issues/142

This PR contains a custom OpenAPI -> DF Marker conversion and implementation in the Gradle- and KSP plugin.
There are also a lot of tests, but I'm not yet 100% confident I've caught every edge case (there are many...).
Docs are updated too.
Let me know if there are any major things (or minor ones, all are welcome) that need changing or if you have some more testing ideas.

koperagen commented on plugins/symbol-processor/src/test/kotlin/org/jetbrains/dataframe/ksp/DataFrameSymbolProcessorTest.kt:
```diff
@@ -888,6 +1043,173 @@ class DataFrameSymbolProcessorTest {
     }
 
+    private val petstoreYaml = File("../../core/src/test/resources/petstore.yaml")
+    private val petstoreYamlUrl = "https://raw.githubusercontent.com/OAI/OpenAPI-Specification/main/examples/v3.0/petstore.yaml"
```
is there a reason why this is downloaded directly from their github instead of using local copy + useHostedFile?

Jolanrensen replied:
I wanted to test whether that would work too, but local might be better

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:
```diff
@@ -0,0 +1,1218 @@
+import kotlin.reflect.KType
+import kotlin.reflect.typeOf
+
+public fun main() {
```
also look like leftovers

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:
```diff
@@ -0,0 +1,1218 @@
+ */
+private fun Map<String, Schema<*>>.toMarkers(): List<MyMarker> {
+    // Convert the schemas to toMarker calls that can be repeated to resolve references.
+    val retrievableMarkers = mapValues { (typeName, value) ->
```
Can you explain this part with references, resolve, retrieve and  ProduceAdditionalMarker? It seems i don't know enough about open api to understand it

Jolanrensen replied:
This is actually something I came up with myself. OpenApi provides schemas for all of the types used. For each type, we want to generate a Marker (Which can be an interface, enum, etc.). However, the OpenApi schema is not ordered per se, so when we are reading the schema it might be that we have a reference to a (super)type (which are queried using `getRefMarker`) for which we have not yet created a Marker. In that case, I "pause" that one (by returning CannotFindRefMarker) and try to read another type schema first. Circular references cannot exist since it's encoded in JSON, so we never get stuck in an infinite loop. When all markers are "retrieved" (so turned from a RetrievableMarker to a MarkerResult.Success), we're done and have converted everything!

As for `produceAdditionalMarker`: In OpenAPI not all enums/objects have to be defined as a separate schema. Although recommended, you can still define an object anonymously directly as a type. For this, I have `produceAdditionalMarker` since during the conversion of a schema->Marker we get an additional new Marker.

Jolanrensen commented on examples/idea-examples/movies/build.gradle.kts:
```diff
@@ -22,3 +22,10 @@ tasks.withType<KotlinCompile> {
     kotlinOptions.jvmTarget = "1.8"
 }
 
+//dataframes {
```
nope XD thanks!

Jolanrensen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/guess.kt:
```diff
@@ -17,17 +18,27 @@ import java.util.ServiceLoader
+    public fun createDefaultReadMethod(pathRepresentation: String? = null): DefaultReadDfMethod
+}
+
+public interface SupportedDataFrameFormat : SupportedFormat {
```
cool! didn't know that

Jolanrensen commented:
okay, I looked over additionalProperties... https://swagger.io/docs/specification/data-models/dictionaries/
Let's see how I can add that

Jolanrensen commented:
Since I changed so much I'm gonna provide a small overview of the changes I made. Might make it easier to review. I'll go through the changed files:

- Linting changes (sorry, but improves readability)
- Docs for OpenAPI and jsonoptions
- replaced all occurrences of "splitted" with "split" because English
- updated kotlinDatetime and made it api() for jupyter etc.
- ImportDataSchema gained jsonOptions with typeClashTactic and keyValuePaths
- starting with kdoc in some functions (more will come later)
- convertTo
  - `convertIf` in dsl: way more powerful than `convert(KType, KType)`, allows you to specify via a condition function whether you want to do a certain conversion and provides fromType and toSchema in ConverterScope. `convert` still has priority over `convertIf`
  - convertTo can now happen for any empty df (both 0 rows or 0 columns)
  - manual converters can now happen for any type of target column, not just value columns. user conversion happens before df conversion making some conversions that were previously impossible possible.
  - Value columns of datarows can become column groups, same as value columns with all nulls.
  - value columns of dataframes can become frame columns
  - absent columns can be created iff it's a nullable typed value column, DataRow<Something?> for a group column or frame column.
- containsNoData() helper function (created that once but don't use it anymore, might still be useful? opinion needed)
- enums can implement `DataSchemaEnum` to control how they are (en/de)coded from/to datasets instead of using just their name
- code generation can now generate enums and type aliases too
- `DefaultReadDfMethod` now provides actual `Marker` instead of just a name
- nullability helpers for FieldType
- bugfix for `MarkersExtractor` regarding `nullableProperties`
- `contains` in `BaseColumn` is now `operator fun`
- Bugfix for concat of DataFrames without columns (rows must still count up, else schema conversions will break)
- String+Number can become Comparable, this was always the case but order dependent. That now always gives the same result thanks to a bugfix in `commonParents()` and `commonType()`
- ISO_DATE_TIME support in parser from string
- `CodeGenerator.Companion.urlReader` is now split into `urlDfReader` and `urlCodeGenReader` where the former is used to generate a dataframe from the data and types from that, while the latter directly generates types (for openapi)
- Similar split for `SupportedFormat` into `SupportedDataFrameFormat` and `SupportedCodeGenerationFormat`
- `createColumn()`
  - bugfix for empty iterables always making column groups. Now uses guesstype or creates value column
  - bugfix for iterables with just nulls always making a frame column, now also uses guesstype or creates value column
  - can now create Column group with iterable of datarows too
- small fix in printing newlines of data schemas
-  `createEmptyColumn` and `createEmptyDataFrame` variants with `numberOfRows`
- `ColumnSchema`, aside from `type` now also has `contentType` for extra type info for Group- and Frame columns (instead of just getting `DataRow<*>` or `DataFrame<*>`. Very useful for new `convertIf` method. `intersectSchema` also tries to merge these `contentType`s if possible. `extractSchema` will make them `Any?`.
- `toSnakeCase()` helper function
- TODO more...

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/TypeUtils.kt:
```diff
@@ -164,20 +175,30 @@ internal fun commonParents(classes: Iterable<KClass<*>>): List<KClass<*>> =
+ *
+ * @param types the types to find the common type for
+ * @param listifyValues if true, then values and nulls will be wrapped in a list if they appear among other lists.
+ *   For example: `[Int, Any?, List<Int>]` will become `List<Int>` instead of `Any?`. If there is another collection
```
is that so? 
It seems like a bug. Because type of null should be Nothing?

Jolanrensen replied:
Definitely not `Nothing?` right? That only appears for thrown errors and return statements. `Nothing?` cannot have an instance. `Int, Any?, List<Int>` will become `Any?` as far as I know.

koperagen replied:
Nothing appears for thrown errors, but `Nothing?` is a special type for null values 
![image](https://user-images.githubusercontent.com/12936457/202169328-c55f7497-f7c4-4a3e-8916-ee6f89c12c31.png)

Jolanrensen replied:
Oh, I see now. I made an error in the KDoc. `[Int, Any?, List<Int>]` becomes `List<Any?>?`.
`[Int, Nothing?, List<Int>]` will become `List<Int>` (or at least, it should, but it doesn't atm... Bug :)  )

Jolanrensen replied:
All handled in latest commit :)

koperagen commented on core/src/test/kotlin/org/jetbrains/kotlinx/dataframe/api/groupBy.kt:
```diff
@@ -9,27 +9,42 @@ class GroupByTests {
     @Test
     fun `groupBy values with nulls`() {
-        val df = dataFrameOf("a", "b")(
+        val df = dataFrameOf(
```
i don't think this file should be reformatted in this commit / PR. It makes it harder to see actual changes. i mean, there are none, right?

Jolanrensen replied:
Unfortunately I have many of these reformats :/ Also for files I've touched but reverted in later commits. So there's no beginning trying to revert them all again I think. We could also view them as one of the "small fixes" part of the PR if that's okay with you :)

koperagen replied:
Sure, let's keep these changes. Just saying for the future

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/CodeGeneratorImpl.kt:
```diff
@@ -62,70 +64,82 @@ internal interface TypeRenderingStrategy {
+                fieldType.typeFqName.shorten()
+
+            is FieldType.GroupFieldType ->
+                if (asAccessor) "$DataRow<${fieldType.markerName}>"
```
Similar conversion happens for FrameFieldType and List. You can generate List<fieldType.markerName> for usages inside data schema declaration and DataFrame<>... for usages in accessors. What do you think about changing this argument to an enum UseSite { DataSchemaDeclaration, GeneratedAccessor }? Or even splitting in two functions. Took me quite a while to understand what's this about

Jolanrensen replied:
2 functions instead of 1 boolean argument is always better indeed :) 
Where is this conversion in lists you talk about?

Jolanrensen replied:
Okay, check out the latest commit, there I split the methods, renamed a bit and added explanation+examples :) should be a lot clearer what's happening now!

koperagen replied:
I mean
@DataSchema
interface Name {
    val firstName: String
}   

@DataSchema
interface Departments {
    val people: List<Name> // Name is DataSchema, so accessor field type is DataFrame. 
    // same as
    val people: DataFrame<Name>
}

Jolanrensen replied:
Hmm, I think this is handled on the schema generation side already. `List<Name>` is a `ValueColumn` type. Can't really seem to find where the distinction happens with wrapping `DataRow` around it or not

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:
```diff
@@ -0,0 +1,1464 @@
+        }
+    ) { (it as DataRow<*>).toMap() }
+
+    // convert DataFrame to DataFrame<Any?> if required by the schema
```
What's the special meaning of DataFrame<Any?>? 
Previously, i think, you could just take any existing dataframe would be a valid value of type DataFrame<Any?>. But here's converting

koperagen replied:
ok, it got commented out :D

Jolanrensen replied:
This is not present anymore in later commits :)

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/ReplCodeGeneratorImpl.kt:
```diff
@@ -68,7 +68,7 @@ internal class ReplCodeGeneratorImpl : ReplCodeGenerator {
                         val requiredBaseMarkers =
                             getRequiredMarkers(columnSchema, registeredMarkers.values)
-                        if (requiredBaseMarkers.all { currentMarker.implements(it) }) {
+                        if (requiredBaseMarkers.any() && requiredBaseMarkers.all { currentMarker.implements(it) }) {
```
is this a fix for some bug? would be nice to have a test to see why this code was changed (or maybe there is? i didn't find it)

Jolanrensen replied:
yup. Accessors wouldn't be generated for my interface-nested DataSchemas without this fix. I think it was just a misinterpretation of `.all {}` again, that's why I didn't write an explicit test here, but the OpenApi Jupyter test breaks if this `.any()` check isn't there.

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/guess.kt:
```diff
@@ -16,18 +19,47 @@ import java.net.URL
+
+    public fun readCodeForGeneration(
+        stream: InputStream,
+        generateHelperCompanionObject: Boolean = false,
```
this in fact needed to support val ApiGuru = importDataSchema() notation in Jupyter, right? I think it's worth mentioning

Jolanrensen replied:
yup!

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/api/KeyValueProperty.kt:
```diff
@@ -0,0 +1,14 @@
+
+/** A [DataSchema] interface / class can implement this if it represents a map-like data schema (so key: value). */
+public interface KeyValueProperty<T> {
+    // needs to be explicitly overridden in @DataSchema interface, otherwise extension functions won't generate (TODO)
```
this interface can be annotated with @DataSchema itself. then extensions going to be a part of the library API and could be called on every marker that implements this interface

Jolanrensen replied:
sure, why not :)

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/convert.kt:
```diff
@@ -76,41 +77,49 @@ internal fun AnyCol.convertToTypeImpl(to: KType): AnyCol {
+
+    if (from.isSubtypeOf(to)) try {
+        return (this as DataColumnInternal<*>).changeType(to.withNullability(hasNulls()))
+    } catch (e: UnsupportedOperationException) { /* */
```
do you remember why?

Jolanrensen replied:
Yes, I believe this fixed a bug where converting from ColumnGroupImpl or FrameColumnImpl to something would crash because it doesn't have the `changeType` option. However, `convertTo` can now support conversions for any type of column, so the exception is caught here to give any user-given conversions a try first, otherwise, a TypeConverterNotFoundException is thrown.

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/convertTo.kt:
```diff
@@ -17,106 +25,209 @@ import org.jetbrains.kotlinx.dataframe.exceptions.ExcessiveColumnsException
 
-    fun getConverter(from: KType, to: KType): Converter? {
-        return converters[from.withNullability(false) to to.withNullability(false)]
+    override fun convertIf(
```
Do you have some representative example for this API? Honestly, i can't easily understand recursive list thing :(

Jolanrensen replied:
For `convertIf`? Sure! It's not that difficult. `convert.with` can only support exact type matches, which can be a problem when you cannot define exact types for everything you want to catch. Let's say you want to write a converter that would convert all numbers to strings:
```kotlin
convert<Number>.with { it.toString() }
```
This wouldn't work, since `typeOf<Double>() != typeOf<Number>()`. That's where `convertIf` can come in handy. With this function you can manufacture any type of converter you like and you can use the full capabilities of Kotlin reflect you like:
```kotlin
convertIf({ fromType, toSchema -> 
    fromType.withNullability(false).isSubtypeOf(typeOf<Number>()) &&
        toSchema.type.withNullability(false) == typeOf<String>()
}) {
    it.toString()
}
```

For OpenApi I needed to provide a converter for `DataRow -> Any`, however, these converters only work on top-level types, so if a `DataRow->Any` conversion needs to happen in a `List<List<List<DataFrame<*>>>`, then we need my recursive function which makes `convertIf` accept any of these types and convert it accordingly.

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:
```diff
@@ -0,0 +1,2173 @@
+}
+
+/** Parse and read OpenApi specification to [DataSchema] interfaces. */
+public fun readOpenApi(
```
why not make DataFrame.Companion receiver?

Jolanrensen replied:
Because you don't create a `DataFrame` ;)

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:
```diff
@@ -0,0 +1,2173 @@
+         * @return a [DataFrame] of [AdditionalProperty]s.
+         */
+        @Suppress("UNCHECKED_CAST")
+        public fun AnyFrame.convertToAdditionalProperties(
```
maybe mention that this used in generated code?

Jolanrensen replied:
sure

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:
```diff
@@ -0,0 +1,2173 @@
+import kotlin.reflect.full.isSubtypeOf
+import kotlin.reflect.jvm.jvmErasure
+import kotlin.reflect.typeOf
+
```
wdyt about splitting this file? also, at some point we started to put formats in different modules

Jolanrensen replied:
Okay splitting off openapi works up 'till the part where I decide to read a file as json or as openapi. Need to think of something clever or universal to make that work elegantly :)

Jolanrensen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/api/DataRowApi.kt:
```diff
@@ -22,6 +23,24 @@ import kotlin.reflect.KType
 public fun AnyRow.isNotEmpty(): Boolean = !isEmpty()
 
+/** Returns `true` even if the columns contain empty iterables, empty dataframes etc. */
+public fun Any?.containsNoData(): Boolean =
```
Okay I agree :) will remove it
